lib/qx/tool/compiler/utils/Json.js

```diff
@@ -17,69 +17,152 @@
  *      * John Spackman (john.spackman@zenesis.com, @johnspackman)
+ *      * oetiker
+ *      * cboulanger
  *
  * *********************************************************************** */
 
 require("@qooxdoo/framework");
+const Ajv = require("ajv");
+const betterAjvErrors = require("better-ajv-errors");
 const fs = qx.tool.compiler.utils.Promisify.fs;
 const JsonToAst = require("json-to-ast");
-
+const path = require("path");
 qx.Class.define("qx.tool.compiler.utils.Json", {
-  
+
   statics: {
     /**
      * Parses JSON string into an object
-     * 
      * @param str {String} the data to parse
      * @return {Object}
+     * @todo similar to qx.tool.cli.commands.Command.parseJsonFile()
      */
     parseJson: function(str) {
-      if (str === null) {
+      if (str === null || !str.trim()) {
         return null;
       }
-      str = str.trim();
-      if (!str) {
+      let ast = JsonToAst.parseToAst(str.trim());
+      return JsonToAst.astToObject(ast);
+    },
+
+    /**
+     * Validates a json object against the given schema signature and outputs
+     * diagnostic information if validation failed
+     * @param json {Object} The json object to check
+     * @param schema {Array|Object}
+     *    The json-schema object or an array of schema objects. If array,
+     *    only the first is used to validate, but the first schema can
+     *    refer to the others.
+     * @param warnOnly {Boolean} If true, do not throw a fatal error
+     * @return {Boolean}
+     *    Returns true if successful and false on failure if the
+     *    'warnOnly' parameter is true
+     */
+    validate(json, schema, warnOnly=false) {
+      let ajv = new Ajv({
+        allErrors: true,
+        jsonPointers: true
+      });
+      if (qx.lang.Type.isArray(schema)) {
+        ajv.addSchema(schema);
+        schema = schema[0].$id;
+      }
+      if (ajv.validate(schema, json)) {
+        // success!
+        return true;
+      }
+      if (warnOnly) {
+        const message = betterAjvErrors(schema.$id, json, ajv.errors, {format: "cli", indent: 2});
+        console.warn("JSON data does not validate against "+ schema.$id + ":\n" + message);
+        return false;
+      }
+      // throw fatal error
+      let err = betterAjvErrors(schema.$id, json, ajv.errors, {format: "js"});
+      let msg;
+      if (err.length) {
+        msg = err[0].error;
+      } else {
+        err = ajv.errors[0];
+        msg = `${err.dataPath} ${err.message}`;
+      }
+      throw new Error(msg);
+    },
+
+    /**
+     * Identify the type and version of the config file schema in the data that
+     * has been passed. Return an object containing information on that schema
+     * which can be used to validate the data.
+     * @param data {Object} Manifest.json or compile.json
+     * @return {{type,version,file,url}|null} Object with information about the
+     * schema or null if no schema could be detected.
+     */
+    getSchemaInfo: function(data) {
```

Comment:
is that still needed?

Reply:
yes, this method provides metadata on the config file, such as type, version, the path to the local copy of the schema file of the schema., I realize that computing the `url` return value in line 131 doesn't seem to  make much sense, since it is in the `$schema` property already. I have to think about whether is should be discarded or if it is a way to check the `$schema` value.

Reply:
I was more wondering about the auto-sensing capability ... I guess that should not be needed anymore as the one reading the config file knows which one is wanted.

Reply:
You are right - there is a logic error here - ATM, we get the schema from the actual data which is loaded into the model and check afterwards if it is the right one whereas the config file model knows which schema it should and should enforce it right away. I'll correct it accordingly.